stdlib/public/core/GraphemeIterator.swift

```diff
@@ -0,0 +1,341 @@
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// GraphemeIterator is a tool to determine the Unicode grapheme break boundaries.
+// This works on strings of UTF8 and UTF16 (for bridged NSString) and returns
+// either the next or previous stride of the grapheme cluster. For example:
+//
+//   let someFlags = "🇺🇸🇬🇧"
+//
+// Byte and Scalar Representation:
+//
+//   |(       🇺       ) (        🇸       )|(        🇬       ) (        🇧      )|
+//   [240, 159, 135, 186, 240, 159, 135, 184, 240, 159, 135, 172, 240, 159, 135, 167]
+//    ^                                        ^
+//    | = String's start index                 | = String's next index
+//
+//
+// In this representation we see that there are a total of 16 utf8 code units,
+// but we don't want to report the count of a string as 16 to a user to sees
+// two emoji "character"s. We also see a total of 4 Unicode scalars, 2 of which
+// make up the single U.S. flag (two regional indicators), i.e. these two create
+// a "grapheme cluster" who represent a singular grapheme or in Swift's terms,
+// "Character". The job of GraphemeIterator is to determine these grapheme
+// cluster boundaries and to correctly report that the string "🇺🇸🇬🇧" has two
+// characters (🇺🇸 and 🇬🇧).
+internal struct GraphemeIterator {
+  let scalars: String.UnicodeScalarView
+
+  var index: String.Index
+  var offset: Int
+
+  var cachedGBP: (Unicode.Scalar, Unicode.GraphemeBreakProperty)? = nil
+
+  var inEmojiSequence: Bool? = nil
+  var shouldBreakRI: Bool? = nil
+
+  init(_ scalars: String.UnicodeScalarView, offset: Int) {
+    self.scalars = scalars
+    self.index = String.Index(_encodedOffset: offset)
+    self.offset = offset
+  }
+
+  mutating func getGraphemeBreakProperty(
+    for scalar: Unicode.Scalar
+  ) -> Unicode.GraphemeBreakProperty {
+    if let cache = cachedGBP, cache.0 == scalar {
+      return cache.1
+    }
+
+    let gbp = Unicode.GraphemeBreakProperty(from: scalar)
+    cachedGBP = (scalar, gbp)
+    return gbp
+  }
+}
+
+
+extension GraphemeIterator {
+  // Returns the stride of the next grapheme cluster at the previous boundary at
+  // offset.
+  mutating func next() -> Int {
+    inEmojiSequence = false
+    shouldBreakRI = false
+
+    while scalars.index(after: index) != scalars.endIndex {
+      let sc1 = scalars[index]
+      scalars.formIndex(after: &index)
+
+      let sc2 = scalars[index]
+
+      let gbp1 = getGraphemeBreakProperty(for: sc1)
+      let gbp2 = getGraphemeBreakProperty(for: sc2)
+
+      if shouldBreak(gbp1, between: gbp2, in: &self) {
+        return index._encodedOffset &- offset
+      }
```

Comment:
Super early feedback: where do the fast paths live? This could be a function taking scalars that checks fast-paths and then falls back to getting the properties.